File: apps/api/src/gridtrace_api/db/models/inspections.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from geoalchemy2 import Geometry
from sqlalchemy import DateTime, Float, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from gridtrace_api.db.base import Base, created_at_column, uuid_pk


class InspectionMission(Base):
    __tablename__ = "inspection_missions"

    id: Mapped[str] = uuid_pk()
    name: Mapped[str] = mapped_column(String, nullable=False)
    region_id: Mapped[str | None] = mapped_column(ForeignKey("regions.id"), nullable=True)
    status: Mapped[str] = mapped_column(String, default="draft")
    scheduled_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    assigned_team_id: Mapped[str | None] = mapped_column(String, nullable=True)
    route_geometry: Mapped[Any | None] = mapped_column(
        Geometry(geometry_type="LINESTRING", srid=4326), nullable=True
    )
    estimated_total_value: Mapped[float] = mapped_column(Float, default=0.0)
    currency: Mapped[str] = mapped_column(String(3), default="EUR")
    created_by: Mapped[str] = mapped_column(String, default="system")
    created_at: Mapped[datetime] = created_at_column()


class InspectionCase(Base):
    __tablename__ = "inspection_cases"

    id: Mapped[str] = uuid_pk()
    mission_id: Mapped[str] = mapped_column(
        ForeignKey("inspection_missions.id"), nullable=False
    )
    customer_id: Mapped[str] = mapped_column(ForeignKey("customers.id"), nullable=False)
    risk_score_id: Mapped[str | None] = mapped_column(
        ForeignKey("risk_scores.id"), nullable=True
    )
    status: Mapped[str] = mapped_column(String, default="queued")
    priority_rank: Mapped[int] = mapped_column(Integer, default=0)
    scheduled_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    assigned_to: Mapped[str | None] = mapped_column(String, nullable=True)
    recommended_action: Mapped[str | None] = mapped_column(String, nullable=True)
    notes: Mapped[str | None] = mapped_column(String, nullable=True)
    created_at: Mapped[datetime] = created_at_column()


class InspectionOutcome(Base):
    __tablename__ = "inspection_outcomes"

    id: Mapped[str] = uuid_pk()
    inspection_case_id: Mapped[str] = mapped_column(
        ForeignKey("inspection_cases.id"), nullable=False
    )
    outcome: Mapped[str] = mapped_column(String, nullable=False)
    confirmed_loss_type: Mapped[str | None] = mapped_column(String, nullable=True)
    estimated_recovered_kwh: Mapped[float | None] = mapped_column(Float, nullable=True)
    estimated_recovered_value: Mapped[float | None] = mapped_column(Float, nullable=True)
    evidence_json: Mapped[dict[str, Any]] = mapped_column(JSONB, default=dict)
    submitted_by: Mapped[str] = mapped_column(String, default="inspector")
    submitted_at: Mapped[datetime] = created_at_column()

File: packages/contracts/src/inspections.ts
import type { LineString } from "./geojson";

export type MissionStatus = "draft" | "planned" | "in_progress" | "completed" | "cancelled";
export type CaseStatus =
  | "queued"
  | "assigned"
  | "scheduled"
  | "in_progress"
  | "resolved"
  | "dismissed";
export type InspectionOutcomeType =
  | "confirmed_theft"
  | "confirmed_meter_fault"
  | "no_issue_found"
  | "inconclusive";

export interface InspectionQueueItem {
  customer_id: string;
  external_ref: string;
  risk_score: number;
  risk_tier: string;
  inspection_priority: number;
  estimated_loss_kwh: number;
  estimated_loss_value: number;
  currency: string;
  recommended_action: string;
  region_name: string | null;
}

export interface InspectionMission {
  id: string;
  name: string;
  region_id: string | null;
  status: MissionStatus;
  scheduled_date: string | null;
  assigned_team_id: string | null;
  route_geometry: LineString | null;
  estimated_total_value: number;
  currency: string;
  created_by: string;
  created_at: string;
  case_count: number;
}

export interface InspectionCase {
  id: string;
  mission_id: string;
  customer_id: string;
  risk_score_id: string | null;
  status: CaseStatus;
  priority_rank: number;
  scheduled_at: string | null;
  assigned_to: string | null;
  recommended_action: string | null;
  notes: string | null;
}

export interface InspectionOutcome {
  id: string;
  inspection_case_id: string;
  outcome: InspectionOutcomeType;
  confirmed_loss_type: string | null;
  estimated_recovered_kwh: number | null;
  estimated_recovered_value: number | null;
  evidence: Record<string, unknown>;
  submitted_by: string;
  submitted_at: string;
}

export interface CreateMissionRequest {
  name: string;
  region_id?: string | null;
  scheduled_date?: string | null;
  assigned_team_id?: string | null;
}

export interface CreateCaseRequest {
  customer_id: string;
  risk_score_id?: string | null;
  recommended_action?: string | null;
  notes?: string | null;
}

export interface UpdateCaseRequest {
  status?: CaseStatus;
  assigned_to?: string | null;
  scheduled_at?: string | null;
  notes?: string | null;
}

export interface SubmitOutcomeRequest {
  outcome: InspectionOutcomeType;
  confirmed_loss_type?: string | null;
  estimated_recovered_kwh?: number | null;
  estimated_recovered_value?: number | null;
  evidence?: Record<string, unknown>;
}

export interface RouteRequest {
  customer_ids: string[];
}

export interface RouteResponse {
  ordered_customer_ids: string[];
  route_geometry: LineString;
  total_distance_km: number;
}
